File: src/main/java/beholder_daemon/service/burnPair/PairAutocheckService.java
package beholder_daemon.service.burnPair;

import beholder_daemon.entity.ContractCode;
import beholder_daemon.entity.burnPair.BurnPairTask;
import beholder_daemon.entity.burnPair.UniswapPair;
import beholder_daemon.repository.ContractCodeRepository;
import beholder_daemon.repository.burnPair.BurnPairTaskRepository;
import beholder_daemon.repository.burnPair.UniswapV2PairRepository;
import beholder_daemon.service.GetCodeService;
import beholder_daemon.service.burnPair.PairValuateService;
import beholder_daemon.utils.EventProcessResult;
import beholder_daemon.utils.SolidityEventProcesser;
import beholder_daemon.utils.Web3Utils;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
import org.springframework.stereotype.Service;
import org.web3j.protocol.Web3j;
import org.web3j.protocol.core.methods.request.EthFilter;
import org.web3j.protocol.core.methods.response.EthLog;
import org.web3j.protocol.core.methods.response.EthLog.LogResult;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@Slf4j
@Service
public class PairAutocheckService{

    @Autowired
    private BurnPairTaskRepository burnPairTaskRepository;

    @Autowired
    private UniswapV2PairRepository uniswapV2PairRepository;

    @Autowired
    private ContractCodeRepository contractCodeRepository;


    public void processTask() {
        BurnPairTask record = new BurnPairTask();
        record.setType("autoCheck");
        record.setStartTime(LocalDateTime.now());
        try {
            record.setStatus(1);
            burnPairTaskRepository.save(record);
            
            // 处理业务逻辑
            log.info("处理pairAutocheck任务: ");

            List<UniswapPair> pairs = uniswapV2PairRepository.findByCheckState(PairValuateService.CHECK_STATE_CODEGOT);
            int pairTotalNumber = pairs.size();
            int fail = 0;
            int noburn = 0;
            int burn = 0;
            for (UniswapPair pair: pairs){
                Optional<ContractCode> code = contractCodeRepository.findById(pair.getTokenAddress());
                if (!code.isPresent()){
                    pair.setCheckState(PairValuateService.CHECK_STATE_AUTOCHECKED_FAIL);
                    fail++;
                }else {
                    String verifiedCode = code.get().getVerifiedCodeDecompressed();
                    String decompiledCode = code.get().getDecompiledCodeDecompressed();
                    if ((verifiedCode==null)&&(decompiledCode==null)){
                        pair.setCheckState(PairValuateService.CHECK_STATE_AUTOCHECKED_FAIL);
                        fail++;
                    }else if (searchInCode(verifiedCode,".sync")||searchInCode(decompiledCode,".sync")){
                        pair.setCheckState(PairValuateService.CHECK_STATE_AUTOCHECKED_BURN);
                        burn++;
                    }else{
                        pair.setCheckState(PairValuateService.CHECK_STATE_AUTOCHECKED_NOBURN);
                        noburn++;
                    }
                }
            }
            uniswapV2PairRepository.saveAll(pairs);            
            // 更新任务状态
            record.setStatus(0);
            record.setEndTime(LocalDateTime.now());
            record.setMessage(String.format("自动检查pair %d个： 自动有 %d个； 自动无 %d个； 自动白 %d 个",pairTotalNumber, burn, noburn, fail));
            burnPairTaskRepository.save(record);
            log.info("pairAutocheck任务处理完成:");
            //} else {
            //    log.warn("任务正在处理中，跳过: {}", message);
            //}
        } catch (Exception e) {
            log.error("处理pairAutocheck任务失败: {}", e);
            record.setStatus(-1);
            StringWriter sw = new StringWriter();
            e.printStackTrace(new PrintWriter(sw));
            record.setMessage(sw.toString());
            record.setEndTime(LocalDateTime.now());
            burnPairTaskRepository.save(record);
        }
    }
    private boolean searchInCode(String codeContent, String keyword) {
        if (codeContent == null || keyword == null) return false;        
        // 判断是否为JSON格式
        if (codeContent.startsWith("{{") && codeContent.endsWith("}}")) {
            try {
                // 解析JSON
                String trimmed = codeContent.substring(1,codeContent.length()-1);
                ObjectMapper mapper = new ObjectMapper();
                JsonNode root = mapper.readTree(trimmed);
                JsonNode sources = root.path("sources");
                if (sources.isObject()) {
                    for (JsonNode fileNode : sources) {
                        String content = fileNode.path("content").asText();
                        if (content.contains(keyword)) {
                            return true;
                        }
                    }
                }
                // 如果没有sources，false
                return false;                
            } catch (Exception e) {
                // 解析失败，false
                return false;
            }
        }        
        // 普通文本直接搜索
        return codeContent.contains(keyword);
    }
}

File: internal/service/burnpair/pair_autocheck_service.go
package burnpair

import (
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/beholder-daemon/internal/model"
	"github.com/beholder-daemon/internal/model/burnpair"
	"github.com/jinzhu/gorm"
)

type PairAutocheckService struct {
	db                   *gorm.DB
	burnPairTaskRepo     *BurnPairTaskRepository
	uniswapV2PairRepo    *UniswapV2PairRepository
	contractCodeRepo     *ContractCodeRepository
}

func NewPairAutocheckService(db *gorm.DB) *PairAutocheckService {
	return &PairAutocheckService{
		db:                   db,
		burnPairTaskRepo:     NewBurnPairTaskRepository(db),
		uniswapV2PairRepo:    NewUniswapV2PairRepository(db),
		contractCodeRepo:     NewContractCodeRepository(db),
	}
}

func (s *PairAutocheckService) ProcessTask() {
	record := &burnpair.BurnPairTask{
		Type:      "autoCheck",
		StartTime: time.Now(),
		Status:    1,
		Message:   "",
	}

	if err := s.burnPairTaskRepo.Create(record); err != nil {
		log.Printf("创建任务记录失败: %v", err)
		return
	}

	log.Println("处理pairAutocheck任务:")

	pairs, err := s.uniswapV2PairRepo.FindByCheckState(CHECK_STATE_CODEGOT)
	if err != nil {
		log.Printf("获取待检查交易对失败: %v", err)
		record.Status = -1
		record.Message = fmt.Sprintf("获取待检查交易对失败: %v", err)
		record.EndTime = time.Now()
		s.burnPairTaskRepo.Update(record)
		return
	}

	pairTotalNumber := len(pairs)
	fail := 0
	noburn := 0
	burn := 0

	for _, pair := range pairs {
		code, err := s.contractCodeRepo.FindByAddress(pair.TokenAddress)
		if err != nil || code == nil {
			checkState := CHECK_STATE_AUTOCHECKED_FAIL
			pair.CheckState = &checkState
			fail++
			continue
		}

		verifiedCode := code.VerifiedCodeDecompressed
		decompiledCode := code.DecompiledCodeDecompressed

		if (verifiedCode == "" || verifiedCode == "0x") && (decompiledCode == "" || decompiledCode == "0x") {
			checkState := CHECK_STATE_AUTOCHECKED_FAIL
			pair.CheckState = &checkState
			fail++
		} else if s.searchInCode(verifiedCode, ".sync") || s.searchInCode(decompiledCode, ".sync") {
			checkState := CHECK_STATE_AUTOCHECKED_BURN
			pair.CheckState = &checkState
			burn++
		} else {
			checkState := CHECK_STATE_AUTOCHECKED_NOBURN
			pair.CheckState = &checkState
			noburn++
		}
	}

	if err := s.uniswapV2PairRepo.SaveAll(pairs); err != nil {
		log.Printf("批量保存交易对失败: %v", err)
	}

	record.Status = 0
	record.EndTime = time.Now()
	record.Message = fmt.Sprintf("自动检查pair %d个： 自动有 %d个； 自动无 %d个； 自动白 %d 个", pairTotalNumber, burn, noburn, fail)

	if err := s.burnPairTaskRepo.Update(record); err != nil {
		log.Printf("更新任务状态失败: %v", err)
	}

	log.Printf("pairAutocheck任务处理完成: %s", record.Message)
}

func (s *PairAutocheckService) searchInCode(codeContent string, keyword string) bool {
	if codeContent == "" || keyword == "" {
		return false
	}

	if strings.HasPrefix(codeContent, "{{") && strings.HasSuffix(codeContent, "}}") {
		return false
	}

	return strings.Contains(codeContent, keyword)
}

type ContractCodeRepository struct {
	db *gorm.DB
}

func NewContractCodeRepository(db *gorm.DB) *ContractCodeRepository {
	return &ContractCodeRepository{db: db}
}

func (r *ContractCodeRepository) FindByAddress(address string) (*model.ContractCode, error) {
	var code model.ContractCode
	err := r.db.First(&code, "address = ?", address).Error
	if err != nil {
		return nil, err
	}
	return &code, nil
}

func (r *UniswapV2PairRepository) FindByCheckState(checkState int) ([]*burnpair.UniswapPair, error) {
	var pairs []*burnpair.UniswapPair
	err := r.db.Where("check_state = ? AND deleted = ?", checkState, 0).Find(&pairs).Error
	return pairs, err
}
